pkg/planner/util/null_misc.go

```diff
@@ -18,76 +18,98 @@ import (
 	"github.com/pingcap/tidb/pkg/expression"
 	"github.com/pingcap/tidb/pkg/parser/ast"
 	"github.com/pingcap/tidb/pkg/planner/context"
+	"github.com/pingcap/tidb/pkg/planner/core/base"
 )
 
-// IsNullRejected check whether a condition is null-rejected
-// A condition would be null-rejected in one of following cases:
-// If it is a predicate containing a reference to an inner table that evaluates to UNKNOWN or FALSE
-// when one of its arguments is NULL.
-// If it is a conjunction containing a null-rejected condition as a conjunct.
-// If it is a disjunction of null-rejected conditions.
-func IsNullRejected(ctx context.PlanContext, schema *expression.Schema, expr expression.Expression) bool {
-	exprCtx := ctx.GetNullRejectCheckExprCtx()
-	expr = expression.PushDownNot(exprCtx, expr)
-	if expression.ContainOuterNot(expr) {
-		return false
+// allConstants checks if only the expression has only constants.
+func allConstants(ctx expression.BuildContext, expr expression.Expression) bool {
+	if expression.MaybeOverOptimized4PlanCache(ctx, []expression.Expression{expr}) {
+		return false // expression contains non-deterministic parameter
 	}
-	sc := ctx.GetSessionVars().StmtCtx
-	for _, cond := range expression.SplitCNFItems(expr) {
-		if isNullRejectedSpecially(ctx, schema, expr) {
-			return true
-		}
-
-		result := expression.EvaluateExprWithNull(exprCtx, schema, cond)
-		x, ok := result.(*expression.Constant)
-		if !ok {
-			continue
-		}
-		if x.Value.IsNull() {
-			return true
-		} else if isTrue, err := x.Value.ToBool(sc.TypeCtxOrDefault()); err == nil && isTrue == 0 {
-			return true
+	switch v := expr.(type) {
+	case *expression.ScalarFunction:
+		for _, arg := range v.GetArgs() {
+			if !allConstants(ctx, arg) {
+				return false
+			}
 		}
+		return true
+	case *expression.Constant:
+		return true
 	}
 	return false
 }
 
-// isNullRejectedSpecially handles some null-rejected cases specially, since the current in
-// EvaluateExprWithNull is too strict for some cases, e.g. #49616.
-func isNullRejectedSpecially(ctx context.PlanContext, schema *expression.Schema, expr expression.Expression) bool {
-	return specialNullRejectedCase1(ctx, schema, expr) // only 1 case now
+// isNullRejectedInList checks null filter for IN list using OR logic.
+// Reason is that null filtering through evaluation by isNullRejectedSimpleExpr
+// has problems with IN list.
```

Comment:
Better to add a case here to explain why we have to treat IN-LIST as OR here. The rest LGTM!

Reply:
So if the in-list contains a column from the non-inner side, then it can never pass the NF check.

Reply:
The explanation for this case is shown above and in the issue. I guess you guys think it is better to add it in the code. Added that to the comment.